File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/FileBasedDataSource.java
package org.cryptomator.domain.usecases.cloud;

import android.content.Context;

import org.cryptomator.util.Optional;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Date;

public class FileBasedDataSource implements DataSource {

	private final File file;

	private FileBasedDataSource(File file) {
		this.file = file;
	}

	public static FileBasedDataSource from(File file) {
		return new FileBasedDataSource(file);
	}

	@Override
	public Optional<Long> size(Context context) {
		return Optional.of(file.length());
	}

	@Override
	public Optional<Date> modifiedDate(Context context) {
		return Optional.of(new Date(file.lastModified()));
	}

	@Override
	public InputStream open(Context context) throws IOException {
		return new FileInputStream(file);
	}

	@Override
	public DataSource decorate(DataSource delegate) {
		return delegate;
	}

	@Override
	public void close() throws IOException {
		// Do nothing
	}
}

File: domain/src/main/java/org/cryptomator/domain/usecases/cloud/FileBasedDataSource.kt
package org.cryptomator.domain.usecases.cloud

import android.content.Context
import org.cryptomator.util.Optional
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.util.Date

class FileBasedDataSource private constructor(private val file: File) : DataSource {

	override fun size(context: Context): Long {
		return file.length()
	}
	override fun modifiedDate(context: Context): Date? {
		return Date(file.lastModified())
	}
	@Throws(IOException::class)
	override fun open(context: Context): InputStream {
		return FileInputStream(file)
	}

	override fun decorate(delegate: DataSource): DataSource {
		return delegate
	}

	@Throws(IOException::class)
	override fun close() {
		// Do nothing
	}

	override fun modifiedDate(context: Context): Optional<Date> {
		return Optional.of(Date(file.lastModified()))
	}

	companion object {

		@JvmStatic
		fun from(file: File): FileBasedDataSource {
			return FileBasedDataSource(file)
		}
	}
}
